File: backend/app/api/datasets.py
from fastapi import APIRouter, UploadFile, File, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from app.core.database import get_db
from app.services.dataset_service import DatasetService

router = APIRouter()


@router.post("/datasets/upload")
async def upload_dataset(file: UploadFile = File(...), db: AsyncSession = Depends(get_db)):
    """Upload a CSV dataset: validate, profile, store to MinIO, and save metadata."""
    service = DatasetService()
    try:
        result = await service.upload_dataset(file, db)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return result


@router.get("/datasets")
async def list_datasets(db: AsyncSession = Depends(get_db)):
    """List all available datasets."""
    service = DatasetService()
    datasets = await service.get_all_datasets(db)
    return datasets


@router.delete("/datasets/{dataset_id}")
async def delete_dataset(dataset_id: str, db: AsyncSession = Depends(get_db)):
    """Delete a dataset specifically, only if no models rely on it."""
    service = DatasetService()
    try:
        result = await service.delete_dataset(dataset_id, db)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return result

File: automl-frontend/src/api/datasets.js
import api from "./axios";

/**
 * Upload a CSV dataset.
 * POST /datasets/upload  (multipart/form-data)
 */
export const uploadDataset = async (file) => {
    const formData = new FormData();
    formData.append("file", file);

    const res = await api.post("/datasets/upload", formData, {
        headers: { "Content-Type": "multipart/form-data" },
    });
    return res.data;
};

export const deleteDataset = async (datasetId) => {
    const res = await api.delete(`/datasets/${datasetId}`);
    return res.data;
};

/**
 * List all datasets.
 * GET /datasets → array of dataset objects
 */
export const getDatasets = async () => {
    const res = await api.get("/datasets");
    return res.data;
};
